sparse/coo.py

```diff
@@ -270,6 +263,29 @@ def __init__(self, coords, data=None, shape=None, has_duplicates=True,
         self.has_duplicates = has_duplicates
         self.sorted = sorted
 
+    @property
+    def shape(self):
+        """
+        The shape of this array.
+
+        Returns
+        -------
+        tuple[int]
+            The shape of this array.
+
+        Examples
+        --------
+        >>> x = np.zeros((10, 10))
+        >>> s = COO.from_numpy(x)
+        >>> s.shape
+        (10, 10)
+        """
+        return self._shape
+
+    @shape.setter
+    def shape(self, value):
+        self._shape = value
```

Comment:
Perhaps we should avoid the property altogether and just have this be a normal attribute?

Reply:
There's no way to inherit it if we do that. And we need `shape` directly for `ndim`, `size`,  and indirectly for`density`.

Reply:
Maybe we just avoid listing it as an abstract method then?

Reply:
Then there's no point to the base class. We would have to remove all of the properties I mentioned.

The whole point of the class is to access the "common" stuff for all types.

Reply:
Or we can make it have an `__init__` method with just the shape.

Reply:
This seems like it might work?  The super class doesn't necessarily need to be functional on its own.

```python
class SparseArray(object):
    @property
    def ndim(self):
        return len(self.shape)

    @property
    def size(self):
        return reduce(operator.mul, self.shape, 1)
  
    @property
    def density(self):
        return self.nnz / self.size

class COO(SparseArray):
    def __init__(...):
        self.shape = ...
```

Reply:
It doesn't make sense in an structural way to be using things undefined in the parent class. However, I made it into an attribute of the parent class. This keeps it as an attribute and still makes sense in the inheritance hierarchy.